core/git.go

```diff
@@ -462,31 +881,253 @@ func (ref *LocalGitRef) Tree(ctx context.Context, srv *dagql.Server, discardGitD
 	return dir, nil
 }
 
-func (ref *LocalGitRef) Commit(ctx context.Context) (string, error) {
-	var commit string
+func (ref *LocalGitRef) Resolve(ctx context.Context) (string, string, error) {
+	if gitutil.IsCommitSHA(ref.Ref) {
+		return ref.Ref, ref.Ref, nil
+	}
+
+	var commit, fullref string
 	err := ref.Repo.mount(ctx, func(src string) error {
-		var err error
-		commit, err = gitCmd(ctx, src, "rev-parse", ref.Ref)
+		git := gitutil.NewGitCLI(gitutil.WithDir(src))
+
+		targetref := ref.Ref
+		out, err := git.Run(ctx, "symbolic-ref", targetref)
+		if err == nil {
+			targetref = strings.TrimSpace(string(out))
+		} else if !strings.Contains(err.Error(), "is not a symbolic ref") {
+			return err
+		}
+		out, err = git.Run(ctx, "show-ref", "--deref", "--head", targetref, targetref+"^{}")
 		if err != nil {
 			return err
 		}
-		return nil
+		commit, fullref, err = parseGitRefOutput(targetref, string(out), " ")
+		return err
 	})
+	if err != nil {
+		return "", "", err
+	}
+	return commit, fullref, nil
+}
+
+func mountKnownHosts(knownHosts string) (string, error) {
+	tempFile, err := os.CreateTemp("", "known_hosts")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temporary known_hosts file: %w", err)
+	}
+
+	_, err = tempFile.WriteString(knownHosts)
+	if err != nil {
+		os.Remove(tempFile.Name())
+		return "", fmt.Errorf("failed to write known_hosts content: %w", err)
+	}
+
+	err = tempFile.Close()
+	if err != nil {
+		os.Remove(tempFile.Name())
+		return "", fmt.Errorf("failed to close temporary known_hosts file: %w", err)
+	}
+
+	return tempFile.Name(), nil
+}
+
+func mountResolv(dns *oci.DNSConfig) (string, error) {
+	src, err := os.Open("/etc/resolv.conf")
 	if err != nil {
 		return "", err
 	}
-	return commit, nil
+	defer src.Close()
+
+	tempFile, err := os.CreateTemp("", "dagger-git-resolv")
+	if err != nil {
+		return "", fmt.Errorf("create resolv.conf override: %w", err)
+	}
+
+	if err := mergeResolv(tempFile, src, dns); err != nil {
+		os.Remove(tempFile.Name())
+		return "", err
+	}
+
+	err = tempFile.Close()
+	if err != nil {
+		os.Remove(tempFile.Name())
+		return "", fmt.Errorf("failed to close temporary resolv.conf file: %w", err)
+	}
+
+	if err := os.Chmod(tempFile.Name(), 0644); err != nil {
+		os.Remove(tempFile.Name())
+		return "", err
+	}
+
+	return tempFile.Name(), nil
+}
+
+func runWithStandardUmaskAndNetOverride(ctx context.Context, cmd *exec.Cmd, hosts, resolv string) error {
+	errCh := make(chan error)
+
+	go func() {
+		defer close(errCh)
+		runtime.LockOSThread()
```

Comment:
I guess this is why our git logic was so good at triggering that go 1.24 bug...

Someday it would be nice to try restoring the mount namespace and unlock here since burning an OS thread every single time git is invoked is kinda crazy, but obviously pre-existing behavior so not high priority.

Reply:
Mmmmm, I *seem* to remember that there's a reason for not doing this - something to do with it being very hard to *safely* call `UnlockOSThread`.

Maybe https://github.com/golang/go/issues/20458? Struggling to remember the exact reason, it could just be cargo-culted.

Reply:
Yeah you have to be super careful because of issues like the one you linked, but afaict since this code only calls the stdlib `os/exec` stuff that's safe here. It is possible though, we do it for our netns stuff during execs for example.

Either way like I said not a big deal for now; the only world in which I can imagine this being meaningful is if the go runtime/scheduler doesn't optimally handle lots of threads getting killed off, which feels plausible but is pure speculation on my part.